### PR TITLE
Additional toolbar visibility options

### PR DESCRIPTION
**Pull request by Ghostamoose**
Originally suggested by `Tonberry#4901` in the TRP Discord.

Currently there's only two visibility options for the TRP toolbar, it's either shown on login or it's not. Here I'm adding the option to show the toolbar only when in-character, with the toolbar disappearing when the player goes out-of-character. Additionally, there's also options to show on login, and hide by default.

Users can still freely toggle the toolbar using the minimap icon or the `/trp3 switch toolbar` slash command.

image

**To do:**
- [x] - Finalize strings
- [x] - Finalize visibility choices

### CODE REVIEW

File: totalRP3/Locales/Locale.lua [hunk 1]
```diff
@@ -1505,6 +1505,11 @@ If you wish to report %s's profile and you cannot target them you will need to o
 	CO_LOCATION_SHOW_OUT_OF_CHARACTER = "Show out of character players",
 	CO_LOCATION_SHOW_OUT_OF_CHARACTER_TT = "If checked, show map pins for all players that are marked as out of character.",
 
+	CO_TOOLBAR_VISIBLITY = "CO_TOOLBAR_VISIBLITY",
+	CO_TOOLBAR_VISIBILITY_HELP = "CO_TOOLBAR_VISIBILITY_HELP",
+	CO_TOOLBAR_VISIBILITY_1 = "Show on login",
+	CO_TOOLBAR_VISIBILITY_2 = "Only show in-character",
+	CO_TOOLBAR_VISIBILITY_3 = "Hide toolbar",
```
Comment (Meorawr) [inline totalRP3/Locales/Locale.lua:1512]:
```lua
CO_TOOLBAR_VISIBILITY = "Toolbar Visibility"
CO_TOOLBAR_VISIBILITY_HELP = "Configures the conditions under which the toolbar will be shown."
CO_TOOLBAR_VISIBILITY_1 = "Always show"
CO_TOOLBAR_VISIBILITY_1 = "Only show in character"
CO_TOOLBAR_VISIBILITY_1 = "Always hidden"
```

Comment (Ghostamoose) [inline totalRP3/Locales/Locale.lua:1512]: wtb strings

File: totalRP3/Modules/Toolbar/Toolbar.lua [hunk 2]
```diff
@@ -304,10 +304,17 @@ local function onStart()
...
-			help = loc.CO_TOOLBAR_SHOW_ON_LOGIN_HELP,
-			configKey = CONFIG_SHOW_ON_LOGIN,
+			inherit = "TRP3_ConfigDropDown",
+			widgetName = "TRP3_ConfigToolbarVisibility",
+			title = loc.CO_TOOLBAR_VISIBILITY,
+			help = loc.CO_TOOLBAR_VISIBILITY_HELP,
+			listContent = {
+				{loc.CO_TOOLBAR_VISIBILITY_1, 1},
+				{loc.CO_TOOLBAR_VISIBILITY_2, 2},
+				{loc.CO_TOOLBAR_VISIBILITY_3, 3}
```
Comment (Meorawr) [inline totalRP3/Modules/Toolbar/Toolbar.lua:314]: Make a local enum in this file and swap to using named constants.

```lua
local ToolbarVisibilityOption = {
    AlwaysShow = 1,
    OnlyShowInCharacter = 2,
    AlwaysHidden = 3,
};
```

File: totalRP3/Modules/Toolbar/Toolbar.lua [hunk 3]
```diff
@@ -358,10 +365,21 @@ local function onStart()
...
+		if toolbarVisibility == 2 and player:IsInCharacter() or toolbarVisibility == 1 then
+			toolbar:Show()
+		else
+			toolbar:Hide()
 		end
 
+		TRP3_API.events.listenToEvent(TRP3_API.events.REGISTER_DATA_UPDATED, function(unitID, _)
+			if unitID ~= Globals.player_id or getConfigValue(CONFIG_TOOLBAR_VISIBILITY) ~= 2 then return end
+
+			if player:IsInCharacter() then toolbar:Show() else toolbar:Hide() end
```
Comment (Meorawr) [inline totalRP3/Modules/Toolbar/Toolbar.lua:381]: Why is this all on one line?

**Review by Meorawr — CHANGES_REQUESTED** [review]
General comment - where's the ~~lamb sauce~~ semicolons?

File: totalRP3/Locales/Locale.lua [hunk 4]
```diff
@@ -1505,6 +1505,11 @@ If you wish to report %s's profile and you cannot target them you will need to o
 	CO_LOCATION_SHOW_OUT_OF_CHARACTER = "Show out of character players",
 	CO_LOCATION_SHOW_OUT_OF_CHARACTER_TT = "If checked, show map pins for all players that are marked as out of character.",
 
+	CO_TOOLBAR_VISIBILITY = "Toolbar Visibility",
+	CO_TOOLBAR_VISIBILITY_HELP = "MEO_PHONE_HOME", --FIXME: don't forget to fix this later
```
Comment (Ghostamoose) [inline totalRP3/Locales/Locale.lua:1509]: elp string needs elp

**Review by Meorawr — CHANGES_REQUESTED** [review]
I'd also prefer the visibility logic be moved to a function.

```lua
local function ShouldShowToolbar()
    local preferredVisibility = TRP3_API.configuration.getValue(CONFIG_TOOLBAR_VISIBILITY);

    if preferredVisibility == ToolbarVisibilityOption.AlwaysHidden then
        return false;
    elseif preferredVisibility == ToolbarVisibilityOption.OnlyShowInCharacter then
        local player = AddOn_TotalRP3.Player.GetCurrentUser();
        return player:IsInCharacter();
    else
        return true;
    end
end
```

Then all those Show/Hide calls can just be shortened to:

````lua
toolbar:SetShown(ShouldShowToolbar());
````

File: totalRP3/Locales/Locale.lua [hunk 5]
```diff
@@ -1505,6 +1505,12 @@ If you wish to report %s's profile and you cannot target them you will need to o
 	CO_LOCATION_SHOW_OUT_OF_CHARACTER = "Show out of character players",
 	CO_LOCATION_SHOW_OUT_OF_CHARACTER_TT = "If checked, show map pins for all players that are marked as out of character.",
 
+	CO_TOOLBAR_VISIBILITY = "Toolbar Visibility",
+	CO_TOOLBAR_VISIBILITY_HELP = "help me",
```
Comment (Meorawr) [inline totalRP3/Locales/Locale.lua:1509]:
```lua
CO_TOOLBAR_VISIBILITY_HELP = "Configures the conditions under which the toolbar should be shown.|n|nThe visibility of the toolbar can be forcefully toggled by running the |cff00ff00/trp switch toolbar|r slash command or by right-clicking the minimap button, which will persist until either logout or a UI reload."
```

**Review by Meorawr — CHANGES_REQUESTED** [review]
I need to think some more on the actual logic changes.

**Review by Meorawr — DISMISSED** [review]
Approving in terms of functional requirements being met and basic testing of the feature in all states proving rather fruitful, though Sol should give it a once-over to see if anything looks amiss.

File: totalRP3/Modules/Toolbar/Toolbar.lua [hunk 6]
```diff
@@ -1,36 +1,95 @@
...
+local ToolbarVisibilityOption = {
+	AlwaysShow = 1,
+	OnlyShowInCharacter = 2,
+	AlwaysHidden = 3,
+};
+
+local ToolbarMixin = {};
+
+function ToolbarMixin:OnLoad()
+	self.manualVisibility = nil;
```
Comment (Solanya) [inline totalRP3/Modules/Toolbar/Toolbar.lua:23]: I'm guessing this should have been forcedVisibility?

**Review by Solanya — APPROVED** [review]

**Pull request merged**